compiler.py

```diff
@@ -0,0 +1,637 @@
+
+    def visit_For(self, node):
+        """This will be called for any* for loop, though not many will be supported because mlog doesn't offer much
+           in the way of iterators or iterable lists.  The two iterables that are supported are range(start, stop, step)
+           and Links (which iterates through various linked blocks)."""
+        # TODO  It would be fairly natural to construe Unit.bind_next as an iterator, and to articulate code that
+        #       cycles through units with the idiom "for Unit in UnitType( flare ):" so it might make sense to offer
+        #       this as another pre-canned for-loop option.  Of course there are other ways of writing this, but
+        #       the same is true of iterating over links
+
+        target = node.target
+        if not isinstance(target, ast.Name):
+            raise CompilerError(ERR_COMPLEX_ASSIGN, node)
+            # TODO: again the problem isn't that "for 12 in range(5):" is *complex*, it's that 12 isn't the right sort of thing to assign values to
+
+        call = node.iter
+        if not isinstance(call, ast.Call):
+            raise CompilerError(ERR_UNSUPPORTED_ITER, node)
+
+        inject = []
+
+        if isinstance(call.func, ast.Attribute) \
+                and call.func.value.id == 'Env' and call.func.attr == 'links':
+            it = REG_IT_FMT.format(call.lineno, call.col_offset)
+            start, end, step = 0, '@links', 1
+            inject.append(f'getlink {target.id} {it}')
+        elif isinstance(call.func, ast.Name) and call.func.id == 'range':
+            it = target.id
+            argv = call.args
+            argc = len(argv)
+            if argc == 1:
+                start, end, step = 0, self.as_value(argv[0]), 1
+            elif argc == 2:
+                start, end, step = *map(self.as_value, argv), 1
+            elif argc == 3:
+                start, end, step = map(self.as_value, argv)
+            else:
+                raise CompilerError(ERR_BAD_ITER_ARGS, node)
+        else:
+            raise CompilerError(ERR_UNSUPPORTED_ITER, node)
+
+        self._ins.append(f'set {it} {start}')
+
+        self._ins.append(f'jump {{}} greaterThanEq {it} {end}')
+        # TODO  if the step in range(start,stop,step) is negative, the relevant test is lessThanEq
+        condition = len(self._ins) - 1
+
+        self._ins.extend(inject)
+        for subnode in node.body:
+            self.visit(subnode)
+
+        self._ins.append(f'op add {it} {it} {step}')
+        self._ins.append(f'jump {condition} always')
+        self._ins[condition] = self._ins[condition].format(len(self._ins))
+
+    def visit_FunctionDef(self, node):
+        """This will be called for any* function definition of the form def fname( args ): body """
+
+        # TODO forbid recursion (or implement it by storing and restoring everything from stack)
+        #      Telos: There's no need to actively forbid recursion, though it may be good to warn about dangers of it.

```

Comment:
Recursion will break as soon as a variable inside a function is used after the recursive call, because that call will overwrite the value. The simplest solution is to forbid it. A better stackless solution is detecting these cases and only erroring on these. The best solution is detecting what needs to be saved to the stack and restored later and doing that.

Reply:
That won't necessarily break anything. If the code was written with the expectation that all variables, including function arguments are global, then this could be working as expected.  E.g. the following psuedo-code function is technically recursive, but still would work fine.  (I know this could be done about as easily with nested while loops.  In general, recursive calls without a stack+namespaces are always equivalent to a combination of global-variable-assignment + while + continue statements, but sometimes that combination could be convenient.)
```python
def find_things_and_blow_them_up( turret ):
    target = find_new_target_for(turret)
    while True:
        if not target:
            find_things_and_blow_them_up( turret )
        turret.shoot_at(target)
```

Reply:
Well, that's bending Python rules, which can throw off users reading Python code that is actually not Python.